### ISSUE TITLE
Bug: Inconsistent behavior of Indexof() and LastIndexOf()

### ISSUE THREAD

**Issue by NightOwl888** [NightOwl888/J2N#62]
We need our own implementations of `IndexOf()` and `LastIndexOf()` on `StringBuilder` because they are not provided in .NET.

The J2N implementations we have account for `Ordinal` and `OrdinalIgnoreCase` of `StringComparison`, but all other options call `ToString()` on the `StringBuilder` and then cascade the call to `string.IndexOf()` and `string.LastIndexOf()`.

However, in the JDK and Apache Harmony, the overload that allows you to specify `startIndex/fromIndex` do not have any exceptions thrown on the bounds and instead correct anything over `Length - 1` to be equivalent to `Length - 1` and anything below zero will return -1. So, this behavior is inconsistent between Java and .NET.

However, we have no way to re-implement `IndexOf()` and `LastIndexOf()` on `String` or `ReadOnlySpan<char>` so the most logical thing to do is to change the `Ordinal` and `OrdinalIgnoreCase` to throw when `Length` is passed so it is consistent with the rest of the .NET behavior. The only downside is that this means that any logic written in Java that is sloppy with the bounds will need to be corrected. However, we could simply provide a note in the API docs that the behavior is inconsistent with Java and consistent with .NET.

Whatever the fix, it will require a breaking behavior change and cannot be addressed without a major version bump.

**Comment by NightOwl888** [comment]
This is technically what we are accomplishing in #171, except our aim is to stop using `StringBuilder` altogether, make a publicly available `OpenStringBuilder` class (#117) and make a public static utility class where external implementations can use the clamping behavior. After we do that, having these extension methods on `System.Text.StringBuilder` is redundant.

The way `System.Text.StringBuilder` manages memory as chunks is a core problem when porting from `java.util.StringBuilder`, which manages memory as a single contiguous chunk. Maintaining extension methods to patch `IndexOf` and `LastIndexOf` is helpful if using `System.Text.StringBuilder`, but performance suffers due to reliance on the indexer of the string builder. Modern TFMs have a `GetChunks()` method which might be able to make a fairly fast `IndexOf` extension method, but it complicates things if for example, the text to match is over multiple chunks. `LastIndexOf` even more cumbersome to implement because the chunks can only be iterated over in a forward direction.

Anyway, once `OpenStringBuilder` is finished this is pretty low priority. But at this point I would argue to keep the clamping behavior in place to match the behavior of `OpenStringBuilder`, which we will be clamping and change any overloads that don't clamp to do so. This makes it less dangerous to port code from Java when the application that is being ported was careless about calculating the inputs, although it will still be necessary to convert from start/end to startIndex/count.
